Formal2024/Workshops/workshop4.lean

```diff
@@ -0,0 +1,260 @@
+  have xs : x ∈ s := xstu.1.1   -- `x ∈ s \ t` means `x ∈ s ∧ x ∉ t`
+  have xnt : x ∉ t := xstu.1.2
+  have xnu : x ∉ u := xstu.2
+  constructor
+  · exact xs
+  intro xtu
+  -- `x ∈ t ∨ x ∈ u`
+  rcases xtu with xt | xu
+  · exact xnt xt
+  . exact xnu xu
+  tada
+
+-- This is quite lengthy. We can use `rintro` to shorten the proof
+
+example : (s \ t) \ u ⊆ s \ (t ∪ u) := by
+  rintro x ⟨⟨xs, xnt⟩, xnu⟩
+  use xs
+  rintro (xt | xu) <;> contradiction -- We often need brackets when using `rintro` on a conjunction
+  tada
+
+-- Two sets are equal when they have the same elements. This is called "extensionality"
+-- You can use it by typing `ext _`
+-- This is the same tactic used to prove functions are equal if they agree on all inputs
+-- When we define a structure with extensionality-like property, we can mark it with `@[ext]`
+-- so that Lean adds an appropriate extensionality lemma to the `ext` tactic.
+-- More on this in the next lectures.
+
+
+example : s ∩ t = t ∩ s := by
+  ext x -- You can give the variable a name by typing it instead of `_`
+  constructor
+  · rintro ⟨xs, xt⟩; exact ⟨xt, xs⟩
+  . rintro ⟨xt, xs⟩; exact ⟨xs, xt⟩
+  tada
+
+-- A similar statement is the theorem `Subset.antisymm`
+
+#check Subset.antisymm
+
+-- Membership to intersection and union can be unfolded using the following two theorems
+-- (although Lean usually does not need it)
+
+#check mem_union
+#check mem_inter_iff
+
+-- Exercises (MIL 4.1)
+
+-- 1
+example : s ∩ (s ∪ t) = s := by
+  sorry
+
+-- 2
+example : s ∪ s ∩ t = s := by
+  sorry
+
+-- 3
+example : s \ t ∪ t = s ∪ t := by
+  sorry
+
+-- 4
+example : s \ t ∪ t \ s = (s ∪ t) \ (s ∩ t) := by
+  sorry
+
+-- 5
+example : s ∩ t ∪ s ∩ u ⊆ s ∩ (t ∪ u) := by
+  sorry
+
+end
+
+/-
+The main difference between sets and types is the predicate `∈`. In fact, in Lean a set coincides
+with its `∈` predicate:
+
+`Set α` is defined as `α → Prop`. If `s : Set α` and `a : α`, then `a ∈ s` is defined as `s a`.
+The set `{ y | P y }` is defined as the function `fun y ↦ P y` and `x ∈ { y | P y }` means `P x`.
+-/
+--
+
+example : 1 ∈ { x : ℕ | x > 0} := by norm_num
+
+-- The empty set `∅` (you get this by typing `\empty`) is the function `fun y ↦ False`
+-- The universe set is the function `fun y ↦ True`
+
+example (x : ℕ) (h : x ∈ (∅ : Set ℕ)) : False :=
+  h
+
+
+example (x : ℕ) : x ∈ (univ : Set ℕ) :=
+  trivial -- `trivial` is the name of the canonical proof of `True`
+
+-- Exercises (MIL 4.1)
+
+section
+
+variable (s t : Set ℕ)
+
+def evenSet : Set ℕ :=
+  { n | Even n }
+
+def oddSet : Set ℕ :=
+  { n | ¬Even n }
+
+#print Prime
+#print Nat.Prime
+
+-- 6
+example : { n | Nat.Prime n } ∩ { n | n > 2 } ⊆ oddSet := by
+  sorry
+
+-- The notation `∀ x ∈ t` means `∀ x, x ∈ t → ... `, and `∃ x ∈ t` means `∃ x, x ∈ t → ... `
+
+-- 7
+example (h₀ : ∀ x ∈ t, ¬Even x) (h₁ : ∀ x ∈ t, Prime x) : ∀ x ∈ s, ¬Even x ∧ Prime x := by
+  sorry
+
+-- 8
+example (h : ∃ x ∈ s, ¬Even x ∧ Prime x) : ∃ x ∈ t, Prime x := by
+  sorry
```

Comment:
since these are all about even, odd and primes, wouldn't it be a good idea to also add some stuff on them as homework?

like proving odd + even = odd, odd + odd = even, even * · = even, ...

Reply:
I feel like they already have a lot of exercises for the session, but we can keep it in mind in case someone is very fast to finish everything

Reply:
we can add this extra homework later tomorrow / after